### ISSUE TITLE
sprintf deprecation on OSX

### ISSUE THREAD

**Issue by PhilipDeegan** [swig/swig#2502]
Hey,

I'm seeing generated code containing `sprintf` with deprecation errors on OSX.
I don't know much about this function but it says to us snprintf.

https://developer.apple.com/forums/thread/714675

```
error: 'sprintf' is deprecated: This function is provided for compatibility reasons only.  
Due to security concerns inherent in the design of sprintf(3), it is highly recommended that 
you use snprintf(3) instead. [-Werror,-Wdeprecated-declarations]
```

I can compile with `-Wno-deprecated-declarations` but I'd rather not.

Suggestions welcome.

**Comment by jschueller** [comment]
using snprintf is a good advice, I guess every sprintf from Lib/*.swg files can be replaced

**Comment by wsfulton** [comment]
sprintf is not deprecated in the C or C++ standards. Happy to accept patches via Github pull requests to replace with snprintf though if functionally the same.

**Comment by ojwb** [comment]
@wsfulton `snprintf` wasn't in C90 though - are you suggesting we can increase the baseline ISO C version we target?  Probably not unreasonable 24 years after it was standardised, but you've not seemed keen to in previous discussions...

Another option for some languages is to use functionality in the language's C API (that's what we already do for PHP and Python for example).  E.g. for perl there's `my_snprintf` (perldoc says it falls back to using `vsprintf` with a buffer overflow check afterwards, but even that's better than using `sprintf` unchecked as currently).

**Comment by wsfulton** [comment]
Oops, I missed that snprintf is C99. There are some embedded C compilers that don't implement C99 and so I would rather keep with the official SWIG aim of generating code that is [easily compiled by virtually every C/C++ compiler and that they can be used on any platform](https://swig.org/Doc4.1/Introduction.html#Introduction_nn13). 

Some of the sprintf could be replaced by a macro using snprintf if C99 available, else fall back to sprintf. Some of the code can be refactored to use target language specific equivalents as @ojwb suggested. Some could be replaced by C++ sstream, yuck! Adding a warnings suppression for OSX + sprintf in swiglabels.swg would easiest and would be my preferred option. We already have a few equivalents for Visual C++ on Windows.

**Comment by ojwb** [comment]
C++ sstream is significantly slower though.

Suppression seems reasonable, at least for the cases where the maximum buffer size can be safely determined.

**Comment by wsfulton** [comment]
Note that Python usage of sprintf was cleaned up in f734c23e8c9b1316f7c242f25e1fda816770ed77 for swig-4.2.

**Comment by ojwb** [comment]
And for PHP in d7ed1a8b97cb0c4e3c27167a85d165990f15ca31.

**Comment by ojwb** [comment]
I've implemented a patch (#2548) which tries to use `snprintf()` where available (with manual overrides in both directions).

This should address this on macos, but I don't have a mac and our CI doesn't have mac builds, so it'd be helpful if @PhilipDeegan (or someone else with a mac) could confirm.

Switching to target language-specific snprintf-like functionality would still be cleaner, but that's likely to take a while to fully achieve even for those target languages which have something suitable.

**Comment by ojwb** [comment]
I've addressed this in the generated code (which solves the problem reported here).

There are still uses of `sprintf` in:

* The source code to swig itself
* Examples and the testsuite
* The manual
* The CCache subdirectory

Maybe the manual should just use `snprintf` instead, so as to demonstrate current best practice and people who copy and paste from the manual are getting safety by default.

The others are only a concern while building/running tests at least.

It's hard to address this well without being able to use variadic macros, but if those are OK then just using `snprintf` likely is.  In this patch I just added two variants for one and two parameters handled by the format string as that's all we use in typemaps, which could be extended to 3, 4, ... but isn't nice.
